File: src/http_bridge/http_bridge_server.cpp
/**
 * HTTP Bridge Server for OpenClaw <-> APP Communication
 * 实现APP期望的HTTP API端点，桥接ROS 2和Android APP
 */

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <boost/beast/core.hpp>
#include <boost/beast/http.hpp>
#include <boost/beast/version.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/json.hpp>
#include <thread>
#include <memory>
#include <string>
#include <map>
#include <mutex>

namespace beast = boost::beast;
namespace http = beast::http;
namespace net = boost::asio;
using tcp = net::ip::tcp;
namespace json = boost::json;

// 全局状态管理
struct SystemState {
    std::string mode = "active";
    std::string main_mode = "ACTIVE";
    bool is_emergency_stop = false;
    bool is_safety_ok = true;
    int error_code = 0;
    std::string error_message = "";

    // 传感器数据
    float motor1_angle = 0.0f;
    float motor2_angle = 0.0f;
    float imu_angle_x = 0.0f;
    float emg_ch1 = 0.0f;
    int heart_rate = 0;
    float motor1_temp = 25.0f;
    float motor2_temp = 25.0f;

    std::mutex mtx;
};

class HttpBridgeNode : public rclcpp::Node {
public:
    HttpBridgeNode() : Node("http_bridge_node") {
        // 订阅CAN数据（从CAN节点接收传感器数据）
        can_rx_sub_ = this->create_subscription<std_msgs::msg::String>(
            "/can_rx", 10,
            std::bind(&HttpBridgeNode::canRxCallback, this, std::placeholders::_1));

        // 发布控制命令（发送到CAN节点）
        can_tx_pub_ = this->create_publisher<std_msgs::msg::String>("/can_tx", 10);

        RCLCPP_INFO(this->get_logger(), "HTTP Bridge Node initialized");
    }

    void publishCommand(const std::string& cmd_json) {
        auto msg = std_msgs::msg::String();
        msg.data = cmd_json;
        can_tx_pub_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Published command: %s", cmd_json.c_str());
    }

    SystemState& getState() { return state_; }

private:
    void canRxCallback(const std_msgs::msg::String::SharedPtr msg) {
        // 解析CAN数据并更新系统状态
        try {
            auto jv = json::parse(msg->data);
            std::lock_guard<std::mutex> lock(state_.mtx);

            if (jv.as_object().contains("motor1_angle"))
                state_.motor1_angle = jv.at("motor1_angle").as_double();
            if (jv.as_object().contains("motor2_angle"))
                state_.motor2_angle = jv.at("motor2_angle").as_double();
            if (jv.as_object().contains("imu_angle_x"))
                state_.imu_angle_x = jv.at("imu_angle_x").as_double();

        } catch (const std::exception& e) {
            RCLCPP_WARN(this->get_logger(), "Failed to parse CAN data: %s", e.what());
        }
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr can_rx_sub_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr can_tx_pub_;
    SystemState state_;
};

// HTTP请求处理器
class HttpSession : public std::enable_shared_from_this<HttpSession> {
public:
    HttpSession(tcp::socket socket, std::shared_ptr<HttpBridgeNode> node)
        : socket_(std::move(socket)), node_(node) {}

    void run() {
        do_read();
    }

private:
    void do_read() {
        auto self = shared_from_this();
        http::async_read(socket_, buffer_, req_,
            [self](beast::error_code ec, std::size_t) {
                if (!ec) self->handle_request();
            });
    }

    void handle_request() {
        auto const bad_request = [](beast::string_view why) {
            http::response<http::string_body> res{http::status::bad_request, 11};
            res.set(http::field::content_type, "application/json");
            res.body() = json::serialize(json::object{{"error", why}});
            res.prepare_payload();
            return res;
        };

        auto const not_found = [](beast::string_view target) {
            http::response<http::string_body> res{http::status::not_found, 11};
            res.set(http::field::content_type, "application/json");
            res.body() = json::serialize(json::object{{"error", "Not found"}, {"path", target}});
            res.prepare_payload();
            return res;
        };

        auto const server_error = [](beast::string_view what) {
            http::response<http::string_body> res{http::status::internal_server_error, 11};
            res.set(http::field::content_type, "application/json");
            res.body() = json::serialize(json::object{{"error", what}});
            res.prepare_payload();
            return res;
        };

        http::response<http::string_body> res;
        res.version(req_.version());
        res.keep_alive(req_.keep_alive());
        res.set(http::field::content_type, "application/json");

        try {
            std::string target = std::string(req_.target());

            // GET /health - 健康检查
            if (req_.method() == http::verb::get && target == "/health") {
                res.result(http::status::ok);
                res.body() = json::serialize(json::object{{"status", "ok"}});
            }
            // GET /status - 获取系统状态和传感器数据
            else if (req_.method() == http::verb::get && target == "/status") {
                auto& state = node_->getState();
                std::lock_guard<std::mutex> lock(state.mtx);

                json::object status_obj{
                    {"timestamp", std::chrono::system_clock::now().time_since_epoch().count()},
                    {"mode", state.mode},
                    {"main_mode", state.main_mode},
                    {"is_emergency_stop", state.is_emergency_stop},
                    {"is_safety_ok", state.is_safety_ok},
                    {"error_code", state.error_code},
                    {"error_message", state.error_message},
                    {"motor1_angle", state.motor1_angle},
                    {"motor2_angle", state.motor2_angle},
                    {"imu_angle_x", state.imu_angle_x},
                    {"emg_ch1", state.emg_ch1},
                    {"heart_rate", state.heart_rate},
                    {"motor1_temp", state.motor1_temp},
                    {"motor2_temp", state.motor2_temp},
                    {"shoulder_angle", state.motor1_angle},
                    {"elbow_angle", state.motor2_angle},
                    {"lateral_position", state.imu_angle_x}
                };

                res.result(http::status::ok);
                res.body() = json::serialize(status_obj);
            }
            // POST /mode - 切换控制模式
            else if (req_.method() == http::verb::post && target == "/mode") {
                auto jv = json::parse(req_.body());
                std::string mode = json::value_to<std::string>(jv.at("mode"));

                {
                    std::lock_guard<std::mutex> lock(node_->getState().mtx);
                    node_->getState().mode = mode;
                }

                // 发送模式切换命令到CAN
                json::object cmd{{"type", "mode"}, {"mode", mode}};
                node_->publishCommand(json::serialize(cmd));

                res.result(http::status::ok);
                res.body() = json::serialize(json::object{{"success", true}, {"mode", mode}});
            }
            // POST /control - 发送控制指令
            else if (req_.method() == http::verb::post && target == "/control") {
                auto jv = json::parse(req_.body());
                json::object cmd{{"type", "control"}};

                if (jv.as_object().contains("shoulder_angle"))
                    cmd["shoulder_angle"] = jv.at("shoulder_angle").as_double();
                if (jv.as_object().contains("elbow_angle"))
                    cmd["elbow_angle"] = jv.at("elbow_angle").as_double();
                if (jv.as_object().contains("lateral_pos"))
                    cmd["lateral_pos"] = jv.at("lateral_pos").as_double();

                node_->publishCommand(json::serialize(cmd));

                res.result(http::status::ok);
                res.body() = json::serialize(json::object{{"success", true}});
            }
            // POST /memory/execute - 执行记忆动作
            else if (req_.method() == http::verb::post && target == "/memory/execute") {
                auto jv = json::parse(req_.body());
                std::string action_id = json::value_to<std::string>(jv.at("action_id"));

                json::object cmd{{"type", "memory"}, {"action_id", action_id}};
                node_->publishCommand(json::serialize(cmd));

                res.result(http::status::ok);
                res.body() = json::serialize(json::object{{"success", true}});
            }
            // POST /memory/stop - 停止记忆动作
            else if (req_.method() == http::verb::post && target == "/memory/stop") {
                json::object cmd{{"type", "stop_memory"}};
                node_->publishCommand(json::serialize(cmd));

                res.result(http::status::ok);
                res.body() = json::serialize(json::object{{"success", true}});
            }
            // POST /api/command - OpenClaw工具调用
            else if (req_.method() == http::verb::post && target == "/api/command") {
                auto jv = json::parse(req_.body());
                std::string tool = json::value_to<std::string>(jv.at("tool"));
                auto params = jv.at("parameters").as_object();

                json::object cmd{{"type", "tool"}, {"tool", tool}, {"parameters", params}};
                node_->publishCommand(json::serialize(cmd));

                res.result(http::status::ok);
                res.body() = json::serialize(json::object{
                    {"success", true},
                    {"result", "Command sent to OpenClaw"}
                });
            }
            else {
                res = not_found(req_.target());
            }

        } catch (const std::exception& e) {
            res = server_error(e.what());
        }

        res.prepare_payload();
        write_response(std::move(res));
    }

    void write_response(http::response<http::string_body>&& res) {
        auto self = shared_from_this();
        res_ = std::make_shared<http::response<http::string_body>>(std::move(res));

        http::async_write(socket_, *res_,
            [self](beast::error_code ec, std::size_t) {
                self->socket_.shutdown(tcp::socket::shutdown_send, ec);
            });
    }

    tcp::socket socket_;
    beast::flat_buffer buffer_;
    http::request<http::string_body> req_;
    std::shared_ptr<http::response<http::string_body>> res_;
    std::shared_ptr<HttpBridgeNode> node_;
};

// HTTP服务器监听器
class HttpListener : public std::enable_shared_from_this<HttpListener> {
public:
    HttpListener(net::io_context& ioc, tcp::endpoint endpoint,
                 std::shared_ptr<HttpBridgeNode> node)
        : ioc_(ioc), acceptor_(ioc), node_(node) {
        beast::error_code ec;

        acceptor_.open(endpoint.protocol(), ec);
        if (ec) throw beast::system_error{ec};

        acceptor_.set_option(net::socket_base::reuse_address(true), ec);
        if (ec) throw beast::system_error{ec};

        acceptor_.bind(endpoint, ec);
        if (ec) throw beast::system_error{ec};

        acceptor_.listen(net::socket_base::max_listen_connections, ec);
        if (ec) throw beast::system_error{ec};
    }

    void run() {
        do_accept();
    }

private:
    void do_accept() {
        acceptor_.async_accept(
            net::make_strand(ioc_),
            [self = shared_from_this()](beast::error_code ec, tcp::socket socket) {
                if (!ec) {
                    std::make_shared<HttpSession>(std::move(socket), self->node_)->run();
                }
                self->do_accept();
            });
    }

    net::io_context& ioc_;
    tcp::acceptor acceptor_;
    std::shared_ptr<HttpBridgeNode> node_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);

    // 创建ROS节点
    auto node = std::make_shared<HttpBridgeNode>();

    // HTTP服务器配置
    auto const address = net::ip::make_address("0.0.0.0");
    auto const port = static_cast<unsigned short>(8081);

    // 创建IO上下文
    net::io_context ioc{1};

    // 创建并启动HTTP监听器
    std::make_shared<HttpListener>(ioc, tcp::endpoint{address, port}, node)->run();

    RCLCPP_INFO(node->get_logger(), "HTTP Bridge Server started on port %d", port);

    // 在单独线程中运行HTTP服务器
    std::thread http_thread([&ioc]() {
        ioc.run();
    });

    // 运行ROS节点
    rclcpp::spin(node);

    // 清理
    ioc.stop();
    http_thread.join();
    rclcpp::shutdown();

    return 0;
}

File: src/http_bridge/http_bridge/http_bridge_server.py
#!/usr/bin/env python3
"""
HTTP Bridge Server for OpenClaw <-> APP Communication
Python实现版本 - 更容易调试和快速迭代
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from flask import Flask, request, jsonify
from threading import Thread, Lock
import json
import time
import subprocess

# 全局状态
class SystemState:
    def __init__(self):
        self.lock = Lock()
        self.mode = "active"
        self.main_mode = "ACTIVE"
        self.is_emergency_stop = False
        self.is_safety_ok = True
        self.error_code = 0
        self.error_message = ""

        # 传感器数据
        self.motor1_angle = 0.0
        self.motor2_angle = 0.0
        self.imu_angle_x = 0.0
        self.emg_ch1 = 0.0
        self.heart_rate = 0
        self.motor1_temp = 25.0
        self.motor2_temp = 25.0
        self.timestamp = int(time.time() * 1000)

state = SystemState()

# ROS 2节点
class HttpBridgeNode(Node):
    def __init__(self):
        super().__init__('http_bridge_node')

        # 订阅CAN数据
        self.can_rx_sub = self.create_subscription(
            String,
            '/can_rx',
            self.can_rx_callback,
            10
        )

        # 发布控制命令
        self.can_tx_pub = self.create_publisher(String, '/can_tx', 10)

        self.get_logger().info('HTTP Bridge Node initialized')

    def can_rx_callback(self, msg):
        """接收CAN数据并更新系统状态"""
        try:
            data = json.loads(msg.data)
            with state.lock:
                state.motor1_angle = data.get('motor1_angle', state.motor1_angle)
                state.motor2_angle = data.get('motor2_angle', state.motor2_angle)
                state.imu_angle_x = data.get('imu_angle_x', state.imu_angle_x)
                state.emg_ch1 = data.get('emg_ch1', state.emg_ch1)
                state.heart_rate = data.get('heart_rate', state.heart_rate)
                state.motor1_temp = data.get('motor1_temp', state.motor1_temp)
                state.motor2_temp = data.get('motor2_temp', state.motor2_temp)
                state.timestamp = int(time.time() * 1000)
        except json.JSONDecodeError as e:
            self.get_logger().warn(f'Failed to parse CAN data: {e}')

    def publish_command(self, cmd_dict):
        """发布控制命令到CAN"""
        msg = String()
        msg.data = json.dumps(cmd_dict)
        self.can_tx_pub.publish(msg)
        self.get_logger().info(f'Published command: {msg.data}')

# Flask应用
app = Flask(__name__)
ros_node = None

@app.route('/health', methods=['GET'])
def health_check():
    """健康检查"""
    return jsonify({"status": "ok"})

@app.route('/status', methods=['GET'])
def get_status():
    """获取系统状态和传感器数据"""
    with state.lock:
        status_data = {
            "timestamp": state.timestamp,
            "mode": state.mode,
            "main_mode": state.main_mode,
            "is_emergency_stop": state.is_emergency_stop,
            "is_safety_ok": state.is_safety_ok,
            "error_code": state.error_code,
            "error_message": state.error_message,
            "motor1_angle": state.motor1_angle,
            "motor2_angle": state.motor2_angle,
            "imu_angle_x": state.imu_angle_x,
            "emg_ch1": state.emg_ch1,
            "heart_rate": state.heart_rate,
            "motor1_temp": state.motor1_temp,
            "motor2_temp": state.motor2_temp,
            # 兼容字段
            "shoulder_angle": state.motor1_angle,
            "elbow_angle": state.motor2_angle,
            "lateral_position": state.imu_angle_x
        }
    return jsonify(status_data)

@app.route('/mode', methods=['POST'])
def set_mode():
    """切换控制模式"""
    data = request.get_json()
    mode = data.get('mode')

    if not mode:
        return jsonify({"error": "Missing mode parameter"}), 400

    with state.lock:
        state.mode = mode

    # 发送模式切换命令到CAN
    ros_node.publish_command({"type": "mode", "mode": mode})

    return jsonify({"success": True, "mode": mode})

@app.route('/control', methods=['POST'])
def send_control():
    """发送控制指令（被动模式）"""
    data = request.get_json()

    cmd = {"type": "control"}
    if 'shoulder_angle' in data:
        cmd['shoulder_angle'] = float(data['shoulder_angle'])
    if 'elbow_angle' in data:
        cmd['elbow_angle'] = float(data['elbow_angle'])
    if 'lateral_pos' in data:
        cmd['lateral_pos'] = float(data['lateral_pos'])

    ros_node.publish_command(cmd)

    return jsonify({"success": True})

@app.route('/memory/execute', methods=['POST'])
def execute_memory():
    """执行记忆动作"""
    data = request.get_json()
    action_id = data.get('action_id')

    if not action_id:
        return jsonify({"error": "Missing action_id parameter"}), 400

    ros_node.publish_command({"type": "memory", "action_id": action_id})

    return jsonify({"success": True})

@app.route('/memory/stop', methods=['POST'])
def stop_memory():
    """停止记忆动作"""
    ros_node.publish_command({"type": "stop_memory"})
    return jsonify({"success": True})

@app.route('/api/command', methods=['POST'])
def api_command():
    """OpenClaw工具调用"""
    data = request.get_json()
    tool = data.get('tool')
    parameters = data.get('parameters', {})

    if not tool:
        return jsonify({"error": "Missing tool parameter"}), 400

    ros_node.publish_command({
        "type": "tool",
        "tool": tool,
        "parameters": parameters
    })

    return jsonify({
        "success": True,
        "result": "Command sent to OpenClaw"
    })

@app.route('/api/sensor/data', methods=['GET'])
def get_sensor_data():
    """获取传感器数据（扩展端点）"""
    return get_status()

@app.route('/api/training/start', methods=['POST'])
def start_training():
    """开始训练会话"""
    data = request.get_json()
    patient_id = data.get('patient_id', 'unknown')
    mode = data.get('mode', 'active')

    session_id = f"session_{int(time.time())}"

    ros_node.publish_command({
        "type": "training_start",
        "patient_id": patient_id,
        "mode": mode,
        "session_id": session_id
    })

    return jsonify({
        "success": True,
        "session_id": session_id
    })

@app.route('/api/stats', methods=['GET'])
def get_stats():
    """获取统计数据"""
    # TODO: 实现统计数据收集
    return jsonify({
        "total_sessions": 0,
        "total_duration": 0,
        "average_score": 0
    })

def send_to_openclaw(message):
    """发送消息到OpenClaw并获取响应"""
    try:
        # 使用 openclaw agent 命令发送消息
        # 对于复杂任务（摄像头、图片分析等），需要更长的超时时间
        result = subprocess.run(
            ['openclaw', 'agent', '--agent', 'main', '--message', message, '--json'],
            capture_output=True,
            text=True,
            timeout=300  # 5分钟超时，适合复杂任务
        )

        if result.returncode == 0:
            # 解析JSON响应
            try:
                response_data = json.loads(result.stdout)
                # 提取AI的回复文本 - 正确的路径是 result.payloads[0].text
                payloads = response_data.get('result', {}).get('payloads', [])
                if payloads and len(payloads) > 0:
                    reply = payloads[0].get('text', '')
                    if reply:
                        ros_node.get_logger().info(f'OpenClaw response: {reply[:100]}...')
                        return reply

                # 如果没有找到文本，记录完整响应用于调试
                ros_node.get_logger().warn(f'No text in OpenClaw response: {response_data}')
                return None

            except json.JSONDecodeError as e:
                ros_node.get_logger().error(f'Failed to parse OpenClaw response: {e}')
                return None
        else:
            ros_node.get_logger().error(f'OpenClaw command failed: {result.stderr}')
            return None

    except subprocess.TimeoutExpired:
        ros_node.get_logger().error('OpenClaw command timeout (5 minutes)')
        return None
    except Exception as e:
        ros_node.get_logger().error(f'Failed to call OpenClaw: {e}')
        return None

@app.route('/message', methods=['POST'])
def handle_message():
    """处理OpenClaw自然语言消息"""
    data = request.get_json()
    message = data.get('message', '')

    if not message:
        return jsonify({
            "success": False,
            "message": "",
            "error": "Missing message parameter"
        }), 400

    # 记录收到的消息
    ros_node.get_logger().info(f'Received message: {message}')

    # 发布消息到ROS话题（供其他节点处理）
    ros_node.publish_command({
        "type": "natural_language",
        "message": message
    })

    # 调用OpenClaw AI处理
    ai_response = send_to_openclaw(message)

    if ai_response:
        return jsonify({
            "success": True,
            "message": ai_response,
            "error": None
        })
    else:
        # 如果OpenClaw调用失败，返回默认响应
        return jsonify({
            "success": True,
            "message": f"收到消息: {message}（OpenClaw暂时不可用）",
            "error": None
        })

def run_flask_app():
    """在单独线程中运行Flask应用"""
    app.run(host='0.0.0.0', port=8081, debug=False, use_reloader=False)

def main(args=None):
    global ros_node

    # 初始化ROS 2
    rclpy.init(args=args)
    ros_node = HttpBridgeNode()

    # 在单独线程中启动Flask服务器
    flask_thread = Thread(target=run_flask_app, daemon=True)
    flask_thread.start()

    ros_node.get_logger().info('HTTP Bridge Server started on http://0.0.0.0:8081')

    try:
        # 运行ROS 2节点
        rclpy.spin(ros_node)
    except KeyboardInterrupt:
        pass
    finally:
        ros_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
